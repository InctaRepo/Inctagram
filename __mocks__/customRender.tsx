/* eslint-disable */
import React, { ReactElement } from 'react'
import { Provider } from 'react-redux'

import { store } from '@/store'
import { render, renderHook, RenderOptions } from '@testing-library/react'

export const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    // <ThemeProvider theme="light">
    //   <TranslationProvider messages={defaultStrings}>
    <Provider store={store}>{children}</Provider>
    //   </TranslationProvider>
    // </ThemeProvider>
  )
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllTheProviders, ...options })

export const customRenderHook = (hook: any, options?: Omit<RenderOptions, 'wrapper'>) => {
  return renderHook(() => hook(), { wrapper: AllTheProviders, ...options })
}
export * from '@testing-library/react'
export * from '@testing-library/user-event'
// eslint-disable-next-line import/export
export { customRender as render }
