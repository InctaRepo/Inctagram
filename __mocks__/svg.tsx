import React, { SVGProps, forwardRef } from 'react'

const SvgrMock = forwardRef<SVGSVGElement, SVGProps<SVGSVGElement>>((props, ref) => (
  <svg ref={ref} {...props} />
))

export const ReactComponent = SvgrMock
export default SvgrMock
