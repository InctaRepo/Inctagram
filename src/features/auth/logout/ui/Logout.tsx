import { useState } from 'react'

import { clsx } from 'clsx'
import { useRouter } from 'next/router'

import { useLogoutMutation } from '@/features/auth/logout/service/logout'
import s from '@/features/auth/logout/ui/logout.module.scss'
import { clearId, clearToken } from '@/features/auth/signIn'
import { LogoutIcon } from '@/shared/assets/icons/LogoutIcon'
import { RouteNames, variantIconLink } from '@/shared/const'
import { getUserEmail, setAuthMeData } from '@/shared/hoc'
import { useAppDispatch, useAppSelector, useTranslate } from '@/shared/hooks'
import { setVariantIcon, sidebarVariantIconSelector } from '@/shared/sidebar'
import { Button } from '@/ui/button'
import { Modal } from '@/ui/modal'
import { Typography } from '@/ui/typography'

export const Logout = () => {
  const dispatch = useAppDispatch()
  const router = useRouter()
  const variantIcon = useAppSelector(sidebarVariantIconSelector)
  const email = useAppSelector(getUserEmail)
  const [openModal, setOpenModal] = useState<boolean>(false)
  const [logoutUser] = useLogoutMutation()

  const { t } = useTranslate()
  const logoutHandler = async () => {
    await logoutUser()
    dispatch(setAuthMeData({ authMeData: { userId: '', username: '', email: '' } }))
    dispatch(clearToken)
    dispatch(clearId)
    dispatch(setVariantIcon(null))
    router.push(RouteNames.SIGN_IN)
    setOpenModal(false)
  }
  const onModalClose = () => {
    setOpenModal(false)
    dispatch(setVariantIcon(null))
  }
  const onClickOpenModal = () => {
    setOpenModal(true)
    dispatch(setVariantIcon(`${RouteNames.LOGOUT}`.slice(1) as variantIconLink))
  }
  const styles = {
    check: clsx(s.linkMenu, `${RouteNames.LOGOUT}`.startsWith('/' + variantIcon) && s.active),
  }

  return (
    <div>
      <div className={styles.check}>
        <Button variant="link" onClick={onClickOpenModal} className={s.btn}>
          <LogoutIcon
            fill={variantIcon === `${RouteNames.LOGOUT}`.slice(1) ? '#397df6' : 'current'}
            className={s.logo}
          />
          <Typography variant="medium14" className={s.text + styles.check}>
            {t.sidebar.logout}
          </Typography>
        </Button>
      </div>
      <Modal
        modalWidth={'md'}
        title={t.sidebar.logout}
        open={openModal}
        actionButtonName={t.profile.yes}
        cancelButtonName={t.profile.no}
        onClose={onModalClose}
        onCancel={onModalClose}
        onAction={logoutHandler}
      >
        <Typography variant={'regular16'}>{t.profile.confirmLogout(email)}</Typography>
      </Modal>
    </div>
  )
}
