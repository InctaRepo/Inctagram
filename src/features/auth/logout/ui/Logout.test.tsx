import React from 'react'

import { render, screen, userEvent, waitFor } from '@/__mocks__/customRender'

import { Logout } from './Logout'

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({ locale: 'en', push: jest.fn() }),
}))
const email = 'test@example.com'
const mockGetUserEmail = jest.fn(() => email)

jest.mock('@/shared/hoc/model/selectors/getUserEmail/getUserEmail', () => ({
  getUserEmail: () => mockGetUserEmail(),
}))

function setup(jsx: React.JSX.Element) {
  return {
    user: userEvent.setup(),
    ...render(jsx),
  }
}
describe('Logout', () => {
  it('renders Logout button', async () => {
    setup(<Logout />)

    const logoutButton = screen.getByRole('button', { name: 'Log Out' })

    await waitFor(() => expect(logoutButton).toBeInTheDocument())
  })

  it('opens the confirmation modal on button click', async () => {
    const { user } = setup(<Logout />)

    await user.click(screen.getByRole('button', { name: 'Log Out' }))

    const confirmationModal = screen.getByText(
      'Are you really want to log out of your account test@example.com?'
    )

    await waitFor(() => expect(confirmationModal).toBeInTheDocument())
  })

  it('calls logoutHandler when confirming logout', async () => {
    const { user } = setup(<Logout />)

    await user.click(screen.getByRole('button', { name: 'Log Out' }))
    await user.click(screen.getByText('Yes'))

    const confirmationModal = screen.queryByText(
      'Are you really want to log out of your account test@example.com?'
    )

    expect(confirmationModal).not.toBeInTheDocument()
  })

  it('closes the confirmation modal on cancel', async () => {
    const { user } = setup(<Logout />)

    await user.click(screen.getByRole('button', { name: 'Log Out' }))
    await user.click(screen.getByText('No'))

    const confirmationModal = screen.queryByText(
      'Are you really want to log out of your account test@example.com?'
    )

    expect(confirmationModal).not.toBeInTheDocument()
  })
  it('snapshot Logout', () => {
    const snapshot = setup(<Logout />)

    waitFor(() => expect(snapshot).toMatchSnapshot())
  })
})
