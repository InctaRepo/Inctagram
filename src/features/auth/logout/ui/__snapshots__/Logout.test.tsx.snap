// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout snapshot Logout 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style=""
  >
    <div>
      <div
        class="linkMenu"
      >
        <button
          class="link  btn"
        >
          <svg
            class="logo"
            fill="current"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              id="Layer 2"
            >
              <g
                id="log-out"
              >
                <path
                  d="M7 6C7.26522 6 7.51957 5.89464 7.70711 5.70711C7.89464 5.51957 8 5.26522 8 5C8 4.73478 7.89464 4.48043 7.70711 4.29289C7.51957 4.10536 7.26522 4 7 4H5C4.73478 4 4.48043 4.10536 4.29289 4.29289C4.10536 4.48043 4 4.73478 4 5V19C4 19.2652 4.10536 19.5196 4.29289 19.7071C4.48043 19.8946 4.73478 20 5 20H7C7.26522 20 7.51957 19.8946 7.70711 19.7071C7.89464 19.5196 8 19.2652 8 19C8 18.7348 7.89464 18.4804 7.70711 18.2929C7.51957 18.1054 7.26522 18 7 18H6V6H7Z"
                  fill="current"
                  id="Vector"
                />
                <path
                  d="M20.82 11.42L18 7.42001C17.8471 7.20442 17.615 7.05815 17.3545 7.01318C17.0941 6.96821 16.8264 7.02819 16.61 7.18001C16.5018 7.2558 16.4098 7.35226 16.3391 7.46382C16.2684 7.57538 16.2206 7.69983 16.1982 7.83C16.1759 7.96016 16.1796 8.09345 16.2091 8.22218C16.2386 8.35091 16.2933 8.47253 16.37 8.58001L18.09 11H10C9.73478 11 9.48043 11.1054 9.29289 11.2929C9.10536 11.4804 9 11.7348 9 12C9 12.2652 9.10536 12.5196 9.29289 12.7071C9.48043 12.8947 9.73478 13 10 13H18L16.2 15.4C16.1212 15.5051 16.0639 15.6246 16.0313 15.7518C15.9987 15.879 15.9915 16.0114 16.01 16.1414C16.0286 16.2714 16.0726 16.3965 16.1395 16.5095C16.2064 16.6225 16.2949 16.7212 16.4 16.8C16.5731 16.9298 16.7836 17 17 17C17.1552 17 17.3084 16.9639 17.4472 16.8944C17.5861 16.825 17.7069 16.7242 17.8 16.6L20.8 12.6C20.9281 12.4309 20.999 12.2254 21.0026 12.0133C21.0062 11.8012 20.9423 11.5934 20.82 11.42Z"
                  fill="current"
                  id="Vector_2"
                />
              </g>
            </g>
          </svg>
          <p
            class="medium14 primary  textlinkMenu"
          >
            Log Out
          </p>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="linkMenu"
    >
      <button
        class="link  btn"
      >
        <svg
          class="logo"
          fill="current"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            id="Layer 2"
          >
            <g
              id="log-out"
            >
              <path
                d="M7 6C7.26522 6 7.51957 5.89464 7.70711 5.70711C7.89464 5.51957 8 5.26522 8 5C8 4.73478 7.89464 4.48043 7.70711 4.29289C7.51957 4.10536 7.26522 4 7 4H5C4.73478 4 4.48043 4.10536 4.29289 4.29289C4.10536 4.48043 4 4.73478 4 5V19C4 19.2652 4.10536 19.5196 4.29289 19.7071C4.48043 19.8946 4.73478 20 5 20H7C7.26522 20 7.51957 19.8946 7.70711 19.7071C7.89464 19.5196 8 19.2652 8 19C8 18.7348 7.89464 18.4804 7.70711 18.2929C7.51957 18.1054 7.26522 18 7 18H6V6H7Z"
                fill="current"
                id="Vector"
              />
              <path
                d="M20.82 11.42L18 7.42001C17.8471 7.20442 17.615 7.05815 17.3545 7.01318C17.0941 6.96821 16.8264 7.02819 16.61 7.18001C16.5018 7.2558 16.4098 7.35226 16.3391 7.46382C16.2684 7.57538 16.2206 7.69983 16.1982 7.83C16.1759 7.96016 16.1796 8.09345 16.2091 8.22218C16.2386 8.35091 16.2933 8.47253 16.37 8.58001L18.09 11H10C9.73478 11 9.48043 11.1054 9.29289 11.2929C9.10536 11.4804 9 11.7348 9 12C9 12.2652 9.10536 12.5196 9.29289 12.7071C9.48043 12.8947 9.73478 13 10 13H18L16.2 15.4C16.1212 15.5051 16.0639 15.6246 16.0313 15.7518C15.9987 15.879 15.9915 16.0114 16.01 16.1414C16.0286 16.2714 16.0726 16.3965 16.1395 16.5095C16.2064 16.6225 16.2949 16.7212 16.4 16.8C16.5731 16.9298 16.7836 17 17 17C17.1552 17 17.3084 16.9639 17.4472 16.8944C17.5861 16.825 17.7069 16.7242 17.8 16.6L20.8 12.6C20.9281 12.4309 20.999 12.2254 21.0026 12.0133C21.0062 11.8012 20.9423 11.5934 20.82 11.42Z"
                fill="current"
                id="Vector_2"
              />
            </g>
          </g>
        </svg>
        <p
          class="medium14 primary  textlinkMenu"
        >
          Log Out
        </p>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
  "user": {
    "clear": [Function],
    "click": [Function],
    "copy": [Function],
    "cut": [Function],
    "dblClick": [Function],
    "deselectOptions": [Function],
    "hover": [Function],
    "keyboard": [Function],
    "paste": [Function],
    "pointer": [Function],
    "selectOptions": [Function],
    "setup": [Function],
    "tab": [Function],
    "tripleClick": [Function],
    "type": [Function],
    "unhover": [Function],
    "upload": [Function],
  },
}
`;
